docs/flip/1009-optimizer-api.md

```diff
@@ -0,0 +1,453 @@
+  (loss, new_model_state), grads = jax.value_and_grad(loss_fn, has_aux=True)(
+      params)
+  updates, new_opt_state = tx_update_fn(grads, opt_state, params)
+  new_params = optax.apply_updates(params, updates)
+  new_variables = variables.copy(dict(params=new_params, **new_model_state))
+  return new_opt_state, new_variables, loss
+
+
+opt_state = tx.init(variables['params'])
+for batch in ds.as_numpy_iterator():
+  opt_state, variables, loss = train_step(
+      opt_state, variables, batch['image'], batch['label'], model.apply,
+      tx.update)
+  print(loss)
+```
+
+Remarks:
+
+- Since `tx.update()` only transforms the gradient, we still need to call
+  `optax.apply_updates()` to apply these transformed gradients to the
+  parameters.
+- Compared with the [previous API], we can now keep the entire `variables`
+  including the `params` as an input and output to the `train_step()`.
+- Splitting `params` from `variables` is still necessary inside the train step
+  because we only want to compute gradients with respect to `params` and not the
+  entire `variables`.
+- We can still log internal optimizer state, such as the learning rate, as long
+  as Optax transformations expose that information in their respective state.
+  For example, `optax.scale_by_schedule()` currently only exposes
+  `opt_state.count` but could easily be extend to also expose the `step_size`.
+  The same is true for internal optimizer states that change over time.
+
+## Multi Optimizer
+[Multi Optimizer]: #multi-optimizer
+
+The [previous API] defined `flax.optim.MultiOptimizer` for processing different
+parts of the parameter tree with different optimizers:
+
+```python
+biases_traversal = flax.optim.ModelParamTraversal(
+    lambda path, _: path.endswith('/bias'))
+not_biases_traversal = flax.optim.ModelParamTraversal(
+    lambda path, _: not path.endswith('/bias'))
+
+optimizer_def = flax.optim.MultiOptimizer(
+    (biases_traversal, flax.optim.GradientDescent(learning_rate=0.1)),
+    (not_biases_traversal, flax.optim.GradientDescent(learning_rate=0.05)),
+)
+```
+
+Note how we first define a traversal that selects parameters based on their
+path (which is the concatenation of module scopes and variable name), and then
+create a `MultiOptimizer` that binds a different optimizer for each of these
+separate traversals.
+
+Optax has recently implemented `optax.masked()` that can be used for specifying
+gradient transformations that only applied to a subset of the gradients:
+
+```python
+def dict_map_with_path(fn, d, path=''):  # a crude "ModelParamTraversal"
+    if isinstance(d, dict):
+        return {k: dict_map_with_path(fn, v, path+'/'+k) for k, v in d.items()}
+    return fn(path, d)
+
+is_bias_mask = dict_map_with_path(
+    lambda path, value: path.endswith('/bias'), variables['params'])
+not_bias_mask = jax.tree_map(lambda x: not x, is_bias_mask)
+
+tx = optax.chain(
+    optax.masked(optax.sgd(learning_rate=0.1), mask=is_bias_mask),
```

Comment:
I wonder if this should be a mask_fn instead or have that option. It seems to break the pattern where you can define the optimizer without knowing the exact variable structure

Reply:
Yeah, I put a similar comment above:
https://github.com/google/flax/pull/1011#discussion_r588426554

@n2cholas @mtthss what are your thoughts on this?

Reply:
This is a good point, I think updating `optax.masked` to accept either a `mask` or `mask_fn` (but not both) is a good idea. `mask_fn` would take a pytree as input and return a `mask`.

It may even be sufficient to change `optax.masked` to only accept a `mask_fn` (if the user really wants to pass in a premade `mask`, they can simply do `mask_fn = lambda: mask`).

Reply:
Opened an [issue](https://github.com/deepmind/optax/issues/57) on the Optax repo for more visibility/discussion.

Reply:
The PR has been merged! The `mask` argument for `optax.masked` can either be a pytree of booleans or a callable that returns such a pytree given the parameters/updates.